sylvia-derive/src/message.rs

```diff
@@ -663,35 +663,62 @@ impl<'a> GlueMessage<'a> {
         let dispatch_arm = quote! {#contract_name :: #contract (msg) =>msg.dispatch(contract, ctx)};
 
         let deserialization_attempts = interfaces.iter().map(|interface| {
-            let ContractMessageAttr { variant, .. } = interface;
-            let var = Ident::new(&variant.to_string().to_case(Case::Snake), variant.span());
-
+            let ContractMessageAttr {
+                module, variant, ..
+            } = interface;
             quote! {
-                let #var = match val.clone().deserialize_into() {
-                    Ok(msg) => return Ok(Self:: #variant (msg)),
-                    Err(err) => err,
+                let msgs = &#module :: #name ::messages();
+                let map = match val.clone() {
+                    #sylvia ::serde_value::Value::Map(map) => map,
+                    _ => panic!("Unexpected msg format!")
+                };
+                if map.len() != 1 {
+                    panic!("Unexpected number of elements received to deserialize!")
+                }
+                let k = match map.into_iter().next() {
+                    Some(k) => k,
+                    None => panic!("Couldn't access the only element of map!")
                 };
+                if let #sylvia ::serde_value::Value::String(k) = k.0 {
```

Comment:
I am pretty sure all of this can be done once per deserialization, not per attempt. It would also avoid cloning the value all the time. If the item is not a string we just return an error.